lib/transport/transport-udp-socket.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2002-2019 Balabit
+ * Copyright (c) 1998-2019 Balázs Scheidler
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "transport-udp-socket.h"
+#include "transport/transport-socket.h"
+#include "scratch-buffers.h"
+#include "str-format.h"
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <errno.h>
+
+typedef struct _LogTransportUDP LogTransportUDP;
+struct _LogTransportUDP
+{
+  LogTransportSocket super;
+  GSockAddr *local_addr;
+};
+
+#if defined(SYSLOG_NG_HAVE_CTRLBUF_IN_MSGHDR)
+static void
+_feed_aux_from_cmsg(LogTransportUDP *self, LogTransportAuxData *aux, struct msghdr *msg)
+{
+  struct cmsghdr *cmsg;
+
+  for (cmsg = CMSG_FIRSTHDR(msg); cmsg != NULL; cmsg = CMSG_NXTHDR(msg, cmsg))
+    {
+
+#ifdef __FreeBSD__
+      if (cmsg->cmsg_level == IPPROTO_IP && cmsg->cmsg_type == IP_RECVDSTADDR)
+        {
+          gint cmsg_header_len = (gchar *) CMSG_DATA(cmsg) - (gchar *) cmsg;
+          struct sockaddr *sa = (struct sockaddr *) CMSG_DATA(cmsg);
+
+          log_transport_aux_data_set_local_addr_ref(aux, g_sockaddr_new(sa, cmsg->cmsg_len - cmsg_header_len));
```

Comment:
Just saw that in the manual there is a `CMSG_LEN` macro too. Could you use that here?

Reply:
Ill check why I wasn't using that. IIRC it wasn't available everywhere.

Reply:
I went back to check my reasons. 

First of all CMSG_LEN() is not doing what you would expect by its name, e.g. it does not give you the size of the payload given the size of an ancilliary block, rather it expects the size of the payload and adds the overhead on top. It is basically a macro to initialize the cmsg_len member of "struct cmsghdr", which contains both the payload and the - aligned - header.

What we need at that point is given cmsg->cmsg_len, give me the length of the payload.  In most ancilliary messages, the payload size is constant so the code would be simpler by assuming the size of the struct being received in the payload. However, this might not work properly in the IP_RECVDSTADDR case, as the payload is of variable length and the size can only be calculated this way.

As a plug, CMSG_LEN is unfortunately not even portable.

To sum it up:
1) CMSG_LEN is not what we need and there's no function/macro that gives us what we need
2) CMSG_LEN is not portable

I could extract a function that does the size calculation, but since it's only used at one place, I don't consider it very important. Let me know if you think otherwise.